lib/tests/test_ecl_grid.cpp

```diff
@@ -0,0 +1,321 @@
+                double y_center = center_of_cell(y_width, j);
+                REQUIRE_THAT(
+                    zc,
+                    WithinAbs(z_width * k + depth_function(x_center, y_center),
+                              1e-4));
+            }
+
+            THEN("The bottom cell corner has z offset by depth function") {
+                double xc, yc, zc;
+
+                ecl_grid_get_cell_corner_xyz3(grid, i, j, k, 4, &xc, &yc, &zc);
+
+                double x_center = center_of_cell(x_width, i);
+                double y_center = center_of_cell(y_width, j);
+                REQUIRE_THAT(zc,
+                             WithinAbs(z_width * (k + 1) +
+                                           depth_function(x_center, y_center),
+                                       1e-4));
+            }
+
+            THEN("The volume should be simple multiplication") {
+                double volume = ecl_grid_get_cell_volume3(grid, i, j, k);
+                REQUIRE_THAT(volume,
+                             WithinAbs(z_width * x_width * y_width, 1e-4));
+            }
+        }
+
+        ecl_grid_free(grid);
+    }
+}
+
+/**
+ * Generates a grid, using the ecl_grid_alloc_GRDECL_kw constructor, with the
+ * x,y,z coordinates of the 0th corner of the ijkth cel is i,j,k. In other
+ * words this is a grid where the corners are on the whole numbers.
+ *
+ * Takes a vector of i,j,k,c,z tuples which changes the corner c of cell at
+ * i,j,k z value.
+ */
+ecl_grid_type *generate_coordkw_grid(
+    int num_x, int num_y, int num_z,
+    vector<tuple<int, int, int, int, double>> z_vector) {
+    ecl_kw_type *coord_kw =
+        ecl_kw_alloc(COORD_KW, ECL_GRID_COORD_SIZE(num_x, num_y), ECL_FLOAT);
+    ecl_kw_type *zcorn_kw = ecl_kw_alloc(
+        ZCORN_KW, ECL_GRID_ZCORN_SIZE(num_x, num_y, num_z), ECL_FLOAT);
+
+    for (int j = 0; j < num_y; j++) {
+        for (int i = 0; i < num_x; i++) {
+            int offset = 6 * (i + j * num_x);
+            ecl_kw_iset_float(coord_kw, offset, i);
+            ecl_kw_iset_float(coord_kw, offset + 1, j);
+            ecl_kw_iset_float(coord_kw, offset + 2, -1);
+
+            ecl_kw_iset_float(coord_kw, offset + 3, i);
+            ecl_kw_iset_float(coord_kw, offset + 4, j);
+            ecl_kw_iset_float(coord_kw, offset + 5, -1);
+
+            for (int k = 0; k < num_z; k++) {
+                for (int c = 0; c < 4; c++) {
+                    int zi1 = ecl_grid_zcorn_index__(num_x, num_y, i, j, k, c);
+                    int zi2 =
+                        ecl_grid_zcorn_index__(num_x, num_y, i, j, k, c + 4);
+
+                    double z1 = k;
+                    double z2 = k + 1;
+
+                    ecl_kw_iset_float(zcorn_kw, zi1, z1);
+                    ecl_kw_iset_float(zcorn_kw, zi2, z2);
+                }
+            }
+        }
+    }
+
+    for (auto &t : z_vector) {
+        int i, j, k, c;
+        double z;
+        tie(i, j, k, c, z) = move(t);
```

Comment:
You're moving objects out of `z_vector`, thus invalidating them. If you ever use any elements of `z_vector` again then it's UB. The correct behaviour here is probably to not `std::move`.